test/test_testing.py

```diff
@@ -744,412 +744,392 @@ def test_trivial_passing_test(self, device):
         expected = {"a": t1, "b": t2}
 
-        for fn in self.get_assert_fns():
+        for fn in assert_fns():
             with self.assertRaisesRegex(AssertionError, r"key\s+'b'"):
                 fn(actual, expected)
 
-    @onlyCPU
-    def test_type_inequality(self, device):
-        actual = torch.empty(2, device=device)
+    def test_type_inequality(self):
+        actual = torch.empty(2)
         expected = actual.tolist()
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             with self.assertRaisesRegex(UsageError, str(type(expected))):
                 fn()
 
-    @onlyCPU
-    def test_unknown_type(self, device):
+    def test_unknown_type(self):
         actual = "0"
         expected = "0"
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             with self.assertRaisesRegex(UsageError, str(type(actual))):
                 fn()
 
-    @onlyCPU
-    def test_numpy(self, device):
-        tensor = torch.rand(2, 2, dtype=torch.float32, device=device)
+    def test_numpy(self):
+        tensor = torch.rand(2, 2, dtype=torch.float32)
         actual = tensor.numpy()
         expected = actual.copy()
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             fn()
 
-    @onlyCPU
-    def test_scalar(self, device):
-        tensor = torch.rand(1, device=device)
+    def test_scalar(self):
+        tensor = torch.rand(1)
         actual = expected = tensor.item()
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             fn()
 
-    @onlyCPU
-    def test_msg_str(self, device):
+    def test_msg_str(self):
         msg = "Custom error message!"
 
-        actual = torch.tensor(1, device=device)
-        expected = torch.tensor(2, device=device)
+        actual = torch.tensor(1)
+        expected = torch.tensor(2)
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             with self.assertRaisesRegex(AssertionError, msg):
                 fn(msg=msg)
 
-    @onlyCPU
-    def test_msg_callable(self, device):
+    def test_msg_callable(self):
         msg = "Custom error message!"
 
         def make_msg(actual, expected, trace):
             return msg
 
-        actual = torch.tensor(1, device=device)
-        expected = torch.tensor(2, device=device)
+        actual = torch.tensor(1)
+        expected = torch.tensor(2)
 
-        for fn in self.assert_fns_with_inputs(actual, expected):
+        for fn in assert_fns_with_inputs(actual, expected):
             with self.assertRaisesRegex(AssertionError, msg):
                 fn(msg=make_msg)
 
 
-instantiate_device_type_tests(TestAsserts, globals())
+def device_combinations():
```

Comment:
This seems like a bit of overkill; just use the deviceCountAtLeast decorator on the test cases:

https://github.com/pytorch/pytorch/blob/866b19e95d1b28fee07161e92692d2cccc788dcd/torch/testing/_internal/common_device_type.py#L738

https://github.com/pytorch/pytorch/blob/866b19e95d1b28fee07161e92692d2cccc788dcd/torch/testing/_internal/common_device_type.py#L34

That can used to create the permutations, if you like

Reply:
See https://github.com/pytorch/pytorch/pull/56365#discussion_r627138093.